Create custom transactor (#2011)
XRPLF/xrpl-dev-portal: content/tutorials/use-specialized-payment-types/create-custom-transactors.md

---
html: create-custom-transactors.html

> parent: use-specialized-payment-types.html

Reviewer: Maybe the parent page should be "Contribute Code" or something in that tree instead?

blurb: Create custom transactors to interact with the XRP Ledger.
labels:

>   - XRP

Reviewer: This feels like entirely the wrong label for this. If there aren't any that fit, you can come up with new ones or just have none.

---
# Create Custom Transactors

> A _transactor_ is code that processes a transaction and modifies the XRP Ledger. Creating custom transactors enables you to add new functionality to `rippled`.

Reviewer: Maybe just add a note for context that actually using this functionality in the shared network requires going through the amendment process? Something to the effect that developing the transactor is only one step in adding new functionality to the protocol.

This tutorial uses the existing `CreateCheck` transactor as an example. You can view the source files here:

- [Header File](https://github.com/XRPLF/rippled/blob/master/src/ripple/app/tx/impl/CreateCheck.h)
- [CPP File](https://github.com/XRPLF/rippled/blob/master/src/ripple/app/tx/impl/CreateCheck.cpp)

## Header File

Create a header file in this format:

```c++
namespace ripple {

class CreateCheck : public Transactor
{
public:
    static constexpr ConsequencesFactoryType ConsequencesFactory{Normal};

    explicit CreateCheck(ApplyContext& ctx) : Transactor(ctx)
    {
    }

    static NotTEC
    preflight(PreflightContext const& ctx);

    static TER
    preclaim(PreclaimContext const& ctx);

    TER
    doApply() override;
};

}  // namespace ripple
```

Initializing the transactor with `ApplyContext` gives it access to:

- The transaction that triggered the transactor.
- A view of the SLE.
- A journal to log errors.

## CPP File

### 1. Add a `preflight` function.

> The `preflight` function checks for errors before accessing the ledger and incurring fees.

Reviewer: Instead of "incurring fees" I suggest providing a little more context about what you're looking for at this stage and why.

```suggestion
The `preflight` function checks for errors before accessing the ledger. It should check for internal inconsistencies in a transaction, so that invalid and incorrectly specified transactions can be rejected with a minimal amount of work.
```

- `PreflightContext` doesn't have a view of the ledger.
- Use bracket notation to retrieve fields from ledgers and transactions:
        
        auto const curExpiration = (*sle*)[~sfExpiration];
        (*sle)[sfBalance] = (*sle)[sfBalance] + reqDelta;
            
    **Note:** The `~` symbol returns an optional type.

- You can view ledger and transaction schemas here:
    - [`LedgerFormats.cpp`](https://github.com/XRPLF/rippled/blob/master/src/ripple/protocol/impl/LedgerFormats.cpp)
    - [`TxFormats.cpp`](https://github.com/XRPLF/rippled/blob/master/src/ripple/protocol/impl/TxFormats.cpp)

```c++
CreateCheck::preflight(PreflightContext const& ctx)
{
    if (!ctx.rules.enabled(featureChecks))

>         return temDISABLED;

Reviewer: The tutorial should probably mention amendments before this point to clarify the usage of `ctx.rules.enabled` and maybe some mention of result codes. You can link [`TER.cpp`](https://github.com/XRPLF/rippled/blob/master/src/ripple/protocol/impl/TER.cpp) or [`TER.h`](https://github.com/XRPLF/rippled/blob/master/src/ripple/protocol/TER.h) which have the canonical lists of transaction result codes.

    }
    if (ctx.tx[sfAccount] == ctx.tx[sfDestination])
    {
        // They wrote a check to themselves.
        JLOG(ctx.j.warn()) << "Malformed transaction: Check to self.";
        return temREDUNDANT;
    }

    {
        STAmount const sendMax{ctx.tx.getFieldAmount(sfSendMax)};
        if (!isLegalNet(sendMax) || sendMax.signum() <= 0)
        {
            JLOG(ctx.j.warn()) << "Malformed transaction: bad sendMax amount: "
                            << sendMax.getFullText();
            return temBAD_AMOUNT;
        }

        if (badCurrency() == sendMax.getCurrency())
        {
            JLOG(ctx.j.warn()) << "Malformed transaction: Bad currency.";
            return temBAD_CURRENCY;
        }
    }

    if (auto const optExpiry = ctx.tx[~sfExpiration])
    {
        if (*optExpiry == 0)
        {
            JLOG(ctx.j.warn()) << "Malformed transaction: bad expiration";
            return temBAD_EXPIRATION;
        }
    }

    return preflight2(ctx);
}
```

### 2. Add a `preclaim` function.

The `preclaim` functions checks for errors that require viewing information on the ledger.

- `PreclaimContext` has a read-only view of the ledger.

> - This function incurs fees.

Reviewer: What fees? How? Who is paying? Please clarify. I suspect the intent is that a transaction that gets to this stage generally gets put into a ledger and burns the transaction cost even if it fails (so, the sender of the transaction pays). Of course, that's not strictly true if they used `fail_hard`, but I'm not sure how that comes in.

The biggest distinction/turning point where a transaction _should_ burn the transaction cost if at all possible is when it's relayed to other servers. I believe that if a transaction gets to `preclaim` then it's probably going to be relayed unless using `fail_hard` but I'm not totally sure. 

I also think the class of result code used may have an effect on that behavior, but I'm not confident about the details. But looking at the sample code, it looks like `preclaim` should return either `tesSUCCESS` (if all went well so far) or a `tec` code  (if something is wrong). Maybe you aren't supposed to return other classes of codes (like `ter` or `tef`) in a `preclaim` function?

    // destination's owner directory.
    if (dstAccountId != account_)
    {
        auto const page = view().dirInsert(
            keylet::ownerDir(dstAccountId),
            checkKeylet,
            describeOwnerDir(dstAccountId));

        JLOG(j_.trace()) << "Adding Check to destination directory "
                        << to_string(checkKeylet.key) << ": "
                        << (page ? "success" : "failure");

        if (!page)
            return tecDIR_FULL;

        sleCheck->setFieldU64(sfDestinationNode, *page);
    }

    {
        auto const page = view().dirInsert(
            keylet::ownerDir(account_),
            checkKeylet,
            describeOwnerDir(account_));

        JLOG(j_.trace()) << "Adding Check to owner directory "
                        << to_string(checkKeylet.key) << ": "
                        << (page ? "success" : "failure");

        if (!page)
            return tecDIR_FULL;

        sleCheck->setFieldU64(sfOwnerNode, *page);
    }
    // If we succeeded, the new entry counts against the creator's reserve.
    adjustOwnerCount(view(), sle, 1, viewJ);
    return tesSUCCESS;
}
```

> ## Additional Functions
> 
> 

Reviewer: I think you should have at least a line of text here explaining this section.

```suggestion
## Additional Functions

You can add more helper functions to your custom transactor as necessary. There are a few special functions that are relevant in special cases.
```

### `calculateBaseFee`

> The `calculateBaseFee` function calculates the XRP fee to run a transaction. The fee is usually inherited from the base transactor, but transactors such as `EscrowFinish` require additional fees.

Reviewer: This phrasing, I think, hides the intent here. Here's an attempt to make it more obvious:

```suggestion
Most transactions inherit the default [reference transaction cost](transaction-cost.html). However, if your transactor needs to define a non-standard transaction cost, you can replace the transactor's `calculateBaseFee` method with a custom one. For example, the following example shows how `EscrowFinish` transactions charge an additional cost on conditional escrows based on the size of the fulfillment:
```

```c++
XRPAmount
EscrowFinish::calculateBaseFee(ReadView const& view, STTx const& tx)
{
    XRPAmount extraFee{0};

    if (auto const fb = tx[~sfFulfillment])
    {
        extraFee += view.fees().base * (32 + (fb->size() / 16));
    }

    return Transactor::calculateBaseFee(view, tx) + extraFee;
}
```

### `makeTxConsequences`

> `rippled` uses a [`TxConsequences`](https://github.com/XRPLF/rippled/blob/master/src/ripple/app/tx/applySteps.h#L41-L44) class to describe the outcome to an account when applying a transaction. It tracks the fee, maximum possible XRP spent, and how many sequences are consumed by the transaction. There are three types of consequences:

Reviewer: Based on what I'm seeing here, I have a feeling the point of this function is to determine when a transaction is [eligible for the queue](https://xrpl.org/transaction-queue.html#queuing-restrictions) and how it affects the queue eligibility of transactions that come after it. I think, relevantly, this needs to be something that can be evaluated using the fields of the transaction itself without further context like the ledger state. So the function should estimate the transaction's max XRP spend, any implications for account sequence numbers and transaction authorization.

FWIW, "sequences" is not a phrasing we use elsewhere in the docs; for consistency, it's probably better to write "sequence numbers" instead.

- **Normal:** The transactor doesn't affect transaction signing and only consumes an XRP fee.
- **Blocker:** The transactor affects transaction signing, preventing valid transactions from queueing behind it.

> - **Custom:** The transactor needs to do additional work to determination consequences.

Reviewer: You should probably clarify that any transaction that can _spend_ XRP (beyond the basic transaction cost) needs a custom `makeTxConsequences`. The [PaymentChannelCreate transaction's example](https://github.com/XRPLF/rippled/blob/aded4a7a925a5822522e6bbbb4adfd3fec39257b/src/ripple/app/tx/impl/PayChan.cpp#L165-L169) is a nice small one. (It can spend up to the `Amount` field of the transaction).

- Payments sending XRP.
- Tickets consuming more than one sequence.
- Transactions that are normal or blockers, depending on flags or fields set.

```c++
SetAccount::makeTxConsequences(PreflightContext const& ctx)
{
    // The SetAccount may be a blocker, but only if it sets or clears
    // specific account flags.
    auto getTxConsequencesCategory = [](STTx const& tx) {
        if (std::uint32_t const uTxFlags = tx.getFlags();
            uTxFlags & (tfRequireAuth | tfOptionalAuth))
            return TxConsequences::blocker;

        if (auto const uSetFlag = tx[~sfSetFlag]; uSetFlag &&
            (*uSetFlag == asfRequireAuth || *uSetFlag == asfDisableMaster ||
             *uSetFlag == asfAccountTxnID))
            return TxConsequences::blocker;

        if (auto const uClearFlag = tx[~sfClearFlag]; uClearFlag &&
            (*uClearFlag == asfRequireAuth || *uClearFlag == asfDisableMaster ||
             *uClearFlag == asfAccountTxnID))
            return TxConsequences::blocker;

        return TxConsequences::normal;
    };

    return TxConsequences{ctx.tx, getTxConsequencesCategory(ctx.tx)};
}
```

## Next Steps

> Submit your transactor to the XRPL `Devnet` for others to test. See: [Contribute Code to the XRP Ledger](contribute-code-flow.html).

Reviewer: It's nonsensical to say, "submit the transactor to the XRPL's `Devnet`". You can't just... submit raw code to a network, let alone Devnet which is a public test network Ripple runs using specifically code that's been merged to the `develop` branch of the XRPLF `rippled` repository.

I think a more reasonable set of next steps include re-compiling the server and testing the transaction in stand-alone mode. If you put the transaction behind an amendment, you'll need to [force-enable the feature using the config file](https://xrpl.org/test-amendments.html#test-amendments).
